src/modules/formBuilder/components/itemEditor/FormItemEditor.tsx

```diff
@@ -0,0 +1,79 @@
+import { Divider, Drawer, Typography } from '@mui/material';
+import React from 'react';
+import { DeepPartial, useForm } from 'react-hook-form';
+import { ErrorState } from '@/modules/errors/util';
+import FormEditorItemProperties from '@/modules/formBuilder/components/itemEditor/FormEditorItemProperties';
+import {
+  FormDefinitionFieldsForEditorFragment,
+  FormItem,
+  ItemType,
+} from '@/types/gqlTypes';
+
+interface FormItemEditorProps {
+  item: FormItem;
+  definition: FormDefinitionFieldsForEditorFragment;
+  saveLoading: boolean;
+  onSave: (item: FormItem, initialLinkId: string) => void;
+  onDiscard: () => void;
+  onClose: () => void;
+  errorState?: ErrorState;
+}
+
+const FormItemEditor: React.FC<FormItemEditorProps> = ({
+  item,
+  definition,
+  onSave,
+  onDiscard,
+  onClose,
+  saveLoading,
+  errorState,
+}) => {
+  // Top-level form state lives here so that we know if it's dirty when attempting to close the drawer
+  const handlers = useForm<DeepPartial<FormItem>>({
```

Comment:
I used `DeepPartial` as the form item state, so that it can support empty values on required fields, for example an empty link ID while you're changing it. This results in a few places where I needed to cast the state into a FormItem, e.g. before submitting it. This means that we need to make sure the Form state has the validations necessary for required fields (I added an example for Link ID), or else we could end up attempting to submit an invalid form, which would cause issues.

I'm curious to hear if anyone has ideas about other patterns to approach this.

Reply:
I think the way you have done it seems pretty good to me. There aren't that many required fields in this form, so it doesn't seem to add a burdensome amount of repetitive code to have each required field need that `rules` prop. I guess the risk is that we may forget to add such rules when adding a new required field to the form, but the risk of that seems low - the form json structure seems pretty stable to me, it seems unlikely we will add a lot more required fields?

Reply:
Okay good point. I agree it seems OK for us to be responsible for the validation there.